src/app/clusters/color-control-server/color-control-server.cpp

```diff
@@ -3006,6 +3103,89 @@ void ColorControlServer::levelControlColorTempChangeCommand(EndpointId endpoint)
 
 #endif // MATTER_DM_PLUGIN_COLOR_CONTROL_SERVER_TEMP
 
+/*
+ * @brief
+ * Utility function used to update a color control attribute which has the quiet reporting quality.
+ * matching the following report conditions:
+ * - At most once per second, or
+ * - At the start of the movement/transition, or
+ * - At the end of the movement/transition, or
+ * - When it changes from null to any other value and vice versa. (Implicit to the QuieterReportingAttribute class)
+ *
+ * The QuietReportAttribute class is updated with the new value and when the report conditions are met,
+ * this function will return MarkAttributeDirty::kIfChanged.
+ * It is expected that the user will use this return value to trigger a reporting mechanism for the attribute with the new value
+ * (Which was updated in the quietReporter)
+ *
+ * @param quietReporter: The QuieterReportingAttribute<TYPE> object for the attribute to update.
+ * @param newValue: Value to update the attribute with
+ * @param isStartOrEndOfTransition: Boolean that indicatse whether the update is occurring at the start or end of a level transition
+ * @return MarkAttributeDirty::kIfChanged when the attribute must be maredk dirty and be reported. MarkAttributeDirty::kNo when it
+ * when it no report is needed.
+ */
+template <typename Q, typename V>
+MarkAttributeDirty ColorControlServer::SetQuietReportAttribute(QuieterReportingAttribute<Q> & quietReporter, V newValue,
+                                                               bool isStartOrEndOfTransition, uint16_t transitionTime)
+{
+    AttributeDirtyState dirtyState;
+    auto now = System::SystemClock().GetMonotonicTimestamp();
+
+    if (isStartOrEndOfTransition)
+    {
+        // At the start or end of the movement/transition we must report
+        auto predicate = [](const typename QuieterReportingAttribute<Q>::SufficientChangePredicateCandidate &) -> bool {
+            return true;
+        };
+        dirtyState = quietReporter.SetValue(newValue, now, predicate);
+    }
+    else
+    {
+        // During transitions, reports should be at most once per second
+
+        // For "infinite" transition, default reports interval to 10s (100 1/10ths of a second )
+        if (transitionTime == MAX_INT16U_VALUE)
+        {
+            transitionTime = 100;
+        }
+
+        // Opt for the longest interval between reports, 1s or (transitionTime / 4).
+        // Since transitionTime is in 1/10th of a second, convert it to ms (x 100), thus * 100/4 -> * 25
+        System::Clock::Milliseconds64 reportInterval = System::Clock::Milliseconds64(std::max(1000, transitionTime * 25));
+        auto predicate                               = quietReporter.GetPredicateForSufficientTimeSinceLastDirty(reportInterval);
+        dirtyState                                   = quietReporter.SetValue(newValue, now, predicate);
+    }
+
+    return (dirtyState == AttributeDirtyState::kMustReport) ? MarkAttributeDirty::kIfChanged : MarkAttributeDirty::kNo;
```

Comment:
This looks wrong.  If kMustReport is true, you want MarkAttributeDirty::kYes, because the IfChanged value will do the wrong thing because it does not know when things were last reported.

Reply:
Filed https://github.com/project-chip/connectedhomeip/issues/34618 to track.